upup/pkg/fi/cloudup/awstasks/autoscalinggroup.go

```diff
@@ -110,9 +112,10 @@ func (e *AutoscalingGroup) Find(c *fi.Context) (*AutoscalingGroup, error) {
 	}
 
 	actual := &AutoscalingGroup{
-		Name:    g.AutoScalingGroupName,
-		MaxSize: g.MaxSize,
-		MinSize: g.MinSize,
+		Name:                g.AutoScalingGroupName,
+		MaxSize:             g.MaxSize,
+		MinSize:             g.MinSize,
+		MaxInstanceLifetime: g.MaxInstanceLifetime,
```

Comment:
So if you try to remove `maxInstanceLifetime` from the IG, the effect won't be applied.

The reason for this may be that this one is not set to `0` if AWS returns `nil`.

Reply:
I did the change on this commit 58b15841e8bb468f1b7d38870013f31322a8c361.
Is there any e2e tests that could help me validate the intended logic?

Reply:
Unfortunately you need to manually test this by creating a cluster and adding/removing the setting.

Reply:
The change in the mentioned commit won't work on its own. You need to feed `fi.Int64(0)` into the task from the model.

Reply:
> You need to feed fi.Int64(0) into the task from the model.

So IIURC @olemarkus, I need to feed `fi.Int64(0)` by default into the model? here --> https://github.com/kubernetes/kops/pull/13892/files#diff-ae2fb14fcb27f448d20d1943a0c3164d3692f812bec7b0d1e3349e93de2bbd2cR545-R547

I thought that setting `fi.Int64(0)` if and only if there is an existing ASG & the `MaxInstanceLifetime`'s spec is set to nil. This is what the aforementioned commit was expected to do.

Reply:
Correct

Reply:
Ok, but it means that everybody using kops in an AWS environment would have all their ASG changed in terraform :/.

Change in fc0c1274031a9abe9c1487dafb60e7597e870975

Reply:
This change breaks e2e tests 
```
Example error: error creating AutoScalingGroup: maxInstanceLifetime must be between 86400 and 31536000 seconds (inclusive).
```
It seems that 0 is only allowed when updating an ASG.

Reply:
Pass nil to create if the value is 0.

Terraform resources will change because of this, yes. But ASGs change on kops updates anyway

Reply:
@olemarkus After introducing the change to pass nil in asg creation, tests are broken.
We pass `MaxInstanceLifetime:0` from spec to model and after creating the ASG we check if there is a change... and there is indeed a change between what we have in aws api and the model.
```
model: MaxInstanceLifetime: 0
api: MaxInstanceLifetime: nil
```

 https://github.com/kubernetes/kops/blob/34490d07cf8bf396a21703f02d8ccf1973c36074/upup/pkg/fi/cloudup/awstasks/autoscalinggroup.go#L101

 wdyt if in ☝🏼 we also use 0 as default value to avoid the difference between model & api?
```go
	if g.MaxInstanceLifetime == nil {
		actual.MaxInstanceLifetime = fi.Int64(0)
	} else {
		actual.MaxInstanceLifetime = g.MaxInstanceLifetime
	}
```

Extract from tests output:
```
I0704 16:09:43.619841 1826014 executor.go:111] Tasks: 85 done / 85 total; 0 can run
Will modify resources:
  AutoscalingGroup/master-us-test-1a.masters.minimal.example.com
        MaxInstanceLifetime      <nil> -> 0

I0704 16:09:43.619960 1826014 featureflag.go:164] FeatureFlag "SpecOverrideFlag"=false
--- FAIL: TestLifecycleMinimalAWS (2.84s)
    lifecycle_integration_test.go:194: running lifecycle test for cluster minimal.example.com
    lifecycle_integration_test.go:205: Target had changes after executing: Will modify resources:
          AutoscalingGroup/master-us-test-1a.masters.minimal.example.com
                MaxInstanceLifetime      <nil> -> 0
```

Reply:
Yeah. So if API returns `nil`, you need to set `actual.MaxInstanceLifetime` to `0`

Reply:
Fixed in 0d7b8d9453c46c4c78f771627efee26aa30620ca